src/System.Diagnostics.Process/src/System.Diagnostics.Process.csproj

```diff
@@ -12,6 +12,9 @@
     <!-- Suppress unused field warnings when using PlatformNotSupportedException stubs -->
     <NoWarn>0649</NoWarn>
   </PropertyGroup>
+  <PropertyGroup Condition=" '$(TargetsLinux)' == 'true'">
+    <DefineConstants>$(DefineConstants);HAVE_PIPE2</DefineConstants>
```

Comment:
Would it be better to take a lock in the C shim instead right before the pipes are created? https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/Native/Unix/System.Native/pal_process.c#L190

It would avoid duplication of the `HAVE_PIPE2` define, and the lock would be taken over smaller scope.

Reply:
> lock would be taken over smaller scope

On rereading pal_process again the lock is over the entire function except for the `access(filename, X_OK)` check. I don't think this meaningfully pushes the lock down, at least enough to be worth having yet another lock.

Reply:
My primary point was the manual duplication of the `HAVE_PIPE2` condition. It will be prone to be out-of-sync with the `HAVE_PIPE2` condition in the C shims. We do not have a similar `HAVE_XXX` condition in any other .csproj file for this reason. All similar conditions are in the C shims.

I agree that the reduced exclusive lock scope is not going to make huge difference.

Reply:
I spent three hours thinking about how to make the lock in native code (It's not trivial because static type variable = new type(); just doesn't exist). I know how to make this lock in native code.

I am not comfortable with doing it w/o the ability to run a test that uses a build including native code compiled with a different set of defines than the production set.

I consider the out-of-sync risk on HAVE_PIPE2 to be low because it's an explicit list in System.Diagnostics.Process.cs and the most likely failure mode is by far a false negative.

Reply:
> how to make the lock in native code 

Just pthread lock, like this:

```
#ifndef HAVE_PIPE2
// ... comment what this lock is for here ...
static pthread_mutex_t process_create_lock = PTHREAD_MUTEX_INITIALIZER;
#endif

...

#ifndef HAVE_PIPE2
    pthread_mutex_unlock(&process_create_lock);
#endif

...

#ifndef HAVE_PIPE2
    pthread_mutex_unlock(&process_create_lock);
#endif
```

Reply:
Excuse me while I go kick my reference material then turn around and replace this pull request.